rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+1. It atomically moves `$cas.m` into `/var/lib/nix`.
+1. It atomically moves `$cas` into `/var/lib/nix`.
+
+Note that to ensure atomicity, `.prepare` and `.stage` need to be on the same filesystem, and either `.stage` or `.daemon` need so be on the same filesystem as the Store.
+
+#### without Store Daemon
+
+Any user with write access to `/var/lib/nix/.stage` and `/var/lib/nix` can install entries. To do so:
+
+1. They prepare entries to be stored in `/var/lib/nix/.stage`, renaming each entry to its `$cas` and `$cas.m`.
+1. They atomically move prepared entries to `/var/lib/nix`, in reverse dependency order, meaning dependencies of an entry are moved first, and `$cas.m` is moved before `$cas`
+
+Note that to ensure atomicity, `.stage` needs to be on the same filesystem as the Store.
+
+Note that when two writers are trying to install the same `$cas` or `$cas.m`, one of them might get an error, but the end result will be the same (as long as the `$cas` is self-valid). So multiple writers can also be on separate hosts, in a trusted setting.
+
+### Verification
+
+A path in the Store is verified by calculating its `$cas` (see above). If the `$cas` doesn't match, the path is moved to `/var/lib/nix/.quarantaine`, where a sysadmin has to investigate.
+
+Any process with write access to `/var/lib/nix` and `/var/lib/nix/.quarantaine` can do this, for example the Store daemon.
+
+### Garbage collection
+
+Garbage collection needs to identify store paths that are not used by anything on any of the systems sharing the same store. Here we propose a simple mechanism for coordination, but any mechanism is acceptable.
+
+- A host with store write access decides to run garbage collection.
+- It checks that `.gc/running_gc` does not exist or contains a very old timestamp, and writes a unique number to `.gc/will_gc`.
+- After waiting long enough to prevent collisions (for example 10 seconds), it reads `.gc/will_gc` and verifies it contains the unique number it wrote.
+- It clears out `.gc/` except for the file `.gc/will_gc` and adds the file `.gc/running_gc` containing the current timestamp.
+- While it waits for other hosts, it checks the Store for `$cas.m` files that don't have a matching `$cas`.
+- Each host's store daemon monitors `.gc/running_gc` at some interval, for example 1 minute.
+- While this file exists, the daemon must record its root `$cas` entries, by creating 0-length files named `.gc/$cas`.
+- The writer waits long enough for all the hosts to record their GC roots, for example 10 minutes.
+- It verifies that `.gc/will_gc` still contains its unique number
+- After the wait period expired, the writer host scans for store paths that are part of the own and other GC roots. Each `$cas` is atomically moved to `.gc` and deleted; `$cas.m` is also deleted.
+- The `$cas.m` files that still don't have their matching `$cas` are removed. Note that when installing, the `$cas.m` will appear shortly before `$cas` since everything is prepared.
```

Comment:
Is it enough for resolving install-vs-GC conflicts, though? A large enough depedency tree being built probably needs more coordination to avoid being GC-ed at the same time as it is deployed.

Reply:
Right, a build could potentially last for days and intermediate entries are required for building. Just touching the dir isn't enough, but the build dependencies could go into auto-roots while a build is proceeding.

Reply:
Yeah, I guess the correct order is per-machine staging, and during a root you create a temporary GC root (maybe in a separate subdir of your namespace!) before starting to try to move the path to main store, and once all builds you are aware of are no longer running (one way or another), you clean up the temporary roots (separating build-time namespaces means that you track each build separately and do not remove another process or another profile permanent root just because it _also_ was your temporary)